core/user_controller/UserController.py

```diff
@@ -1,6 +1,225 @@
+        if not isinstance(requesting_user, User):
+            raise ValueError("Invalid requesting_user: must be a valid User instance")
 
+        user = get_object_or_404(User, id=user_id)
+        course_ids = UserController._get_course_ids_based_on_role(user)
+        courses = Course.objects.filter(id__in=course_ids)
+        course_overviews = UserController._construct_course_overviews(courses, user)
 
-class UserController:
-    pass # TODO: Create static controller class
\ No newline at end of file
+        return UserController._create_user_profile(user, requesting_user, course_overviews)
+
+    @staticmethod
+    def _get_course_ids_based_on_role(user):
+        if user.role == "Instructor":
+            return user.coursesection_set.values_list("course", flat=True)
+        elif user.role == "TA":
+            return user.tacourseassignment_set.values_list("course", flat=True)
+        return []
+
+    @staticmethod
+    def _create_user_profile(user, requesting_user, course_overviews):
+        if requesting_user.role == 'Admin' or user == requesting_user:
+            return PrivateUserProfile(
+                name=f"{user.first_name} {user.last_name}".strip(),
+                email=user.email,
+                role=user.role,
+                office_hours=user.office_hours,
+                courses_assigned=course_overviews,
+                address=user.address,
+                phone=user.phone
+            )
+        return UserProfile(
+            name=f"{user.first_name} {user.last_name}".strip(),
+            email=user.email,
+            role=user.role,
+            office_hours=user.office_hours,
+            courses_assigned=course_overviews
+        )
+
+    @staticmethod
+    def _construct_course_overviews(courses, user):
+        course_overviews = []
+        for course in courses:
+            course_sections = CourseSection.objects.filter(course=course, instructor=user)
+            lab_sections = LabSection.objects.filter(course=course)
+
+            course_overviews.append(CourseOverview(
+                code=course.course_code,
+                name=course.course_name,
+                semester=course.semester,
+                course_sections=UserController._get_course_section_refs(course_sections),
+                lab_sections=UserController._get_lab_section_refs(lab_sections)
+            ))
+
+        return course_overviews
+
+    @staticmethod
+    def _get_course_section_refs(course_sections):
+        return [
+            CourseSectionRef(
+                section_number=str(s.course_section_number),
+                instructor=UserRef(name=s.instructor.get_full_name(), username=s.instructor.username)
+            )
+            for s in course_sections
+        ]
+
+    @staticmethod
+    def _get_lab_section_refs(lab_sections):
+        return [
+            LabSectionRef(
+                section_number=str(ls.lab_section_number),
+                instructor=(
```

Comment:
It should only include lab sections if the current user is the instructor. This requires looking in the TALabAssignment table because LabSection's don't have instructors (The one part of our DB I would change if i could)